src/NServiceBus.NHibernate/RepeatedFailuresOverTimeCircuitBreaker.cs

```diff
@@ -0,0 +1,68 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading;
+    using Logging;
+
+    class RepeatedFailuresOverTimeCircuitBreaker : IDisposable
+    {
+        public RepeatedFailuresOverTimeCircuitBreaker(string name, TimeSpan timeToWaitBeforeTriggering, Action<Exception> triggerAction)
+        {
+            this.name = name;
+            this.triggerAction = triggerAction;
+            this.timeToWaitBeforeTriggering = timeToWaitBeforeTriggering;
+
+            timer = new Timer(CircuitBreakerTriggered);
+        }
+
+        public void Success()
+        {
+            var oldValue = Interlocked.Exchange(ref failureCount, 0);
+
+            if (oldValue == 0)
+            {
+                return;
+            }
+
+            timer.Change(System.Threading.Timeout.Infinite, System.Threading.Timeout.Infinite);
+            Logger.InfoFormat("The circuit breaker for {0} is now disarmed", name);
+        }
+
+        public void Failure(Exception exception)
+        {
+            lastException = exception;
+            var newValue = Interlocked.Increment(ref failureCount);
+
+            if (newValue == 1)
+            {
+                timer.Change(timeToWaitBeforeTriggering, NoPeriodicTriggering);
+                Logger.WarnFormat("The circuit breaker for {0} is now in the armed state", name);
+            }
+        }
+
+        public void Dispose()
+        {
+            //Injected
+        }
+
+        void CircuitBreakerTriggered(object state)
+        {
+            if (Interlocked.Read(ref failureCount) > 0)
+            {
+                Logger.WarnFormat("The circuit breaker for {0} will now be triggered", name);
+                triggerAction(lastException);
+            }
+        }
+
+        long failureCount;
+        Exception lastException;
```

Comment:
Is the last exception enough? or should we go with an AggregateException instead? That'd require a bit more code to ensure exceptions are aggregated correctly and cleared when successful.